cli/js/40_jupyter.js

```diff
@@ -1,8 +1,339 @@
 const internals = globalThis.__bootstrap.internals;
 
+const $display = Symbol.for("Jupyter.display");
+
+/** Duck typing our way to common visualization and tabular libraries */
+/** Vegalite */
+function isVegaLike(obj) {
+  return obj !== null && typeof obj === "object" && "toSpec" in obj;
+}
+function extractVega(obj) {
+  const spec = obj.toSpec();
+  if (!("$schema" in spec)) {
+    return null;
+  }
+  if (typeof spec !== "object") {
+    return null;
+  }
+  let mediaType = "application/vnd.vega.v5+json";
+  if (spec.$schema === "https://vega.github.io/schema/vega-lite/v4.json") {
+    mediaType = "application/vnd.vegalite.v4+json";
+  } else if (
+    spec.$schema === "https://vega.github.io/schema/vega-lite/v5.json"
+  ) {
+    mediaType = "application/vnd.vegalite.v5+json";
+  }
+  return {
+    [mediaType]: spec,
+  };
+}
+/** Polars */
+function isDataFrameLike(obj) {
+  const isObject = obj !== null && typeof obj === "object";
+  if (!isObject) {
+    return false;
+  }
+  const df = obj;
+  return df.schema !== void 0 && typeof df.schema === "object" &&
+    df.head !== void 0 && typeof df.head === "function" &&
+    df.toRecords !== void 0 && typeof df.toRecords === "function";
+}
+/**
+ * Map Polars DataType to JSON Schema data types.
+ * @param dataType - The Polars DataType.
+ * @returns The corresponding JSON Schema data type.
+ */
+function mapPolarsTypeToJSONSchema(colType) {
+  const typeMapping = {
+    Null: "null",
+    Bool: "boolean",
+    Int8: "integer",
+    Int16: "integer",
+    Int32: "integer",
+    Int64: "integer",
+    UInt8: "integer",
+    UInt16: "integer",
+    UInt32: "integer",
+    UInt64: "integer",
+    Float32: "number",
+    Float64: "number",
+    Date: "string",
+    Datetime: "string",
+    Utf8: "string",
+    Categorical: "string",
+    List: "array",
+    Struct: "object",
+  };
+  // These colTypes are weird. When you console.dir or console.log them
+  // they show a `DataType` field, however you can't access it directly until you
+  // convert it to JSON
+  const dataType = colType.toJSON()["DataType"];
+  return typeMapping[dataType] || "string";
+}
+function extractDataFrame(df) {
+  const fields = [];
+  const schema = {
+    fields,
+  };
+  let data = [];
+  // Convert DataFrame schema to Tabular DataResource schema
+  for (const [colName, colType] of Object.entries(df.schema)) {
+    const dataType = mapPolarsTypeToJSONSchema(colType);
+    schema.fields.push({
+      name: colName,
+      type: dataType,
+    });
+  }
+  // Convert DataFrame data to row-oriented JSON
+  //
+  // TODO(rgbkrk): Determine how to get the polars format max rows
+  //       Since pl.setTblRows just sets env var POLARS_FMT_MAX_ROWS,
+  //       we probably just have to pick a number for now.
+  //
+  data = df.head(50).toRecords();
+  let htmlTable = "<table>";
+  htmlTable += "<thead><tr>";
+  schema.fields.forEach((field) => {
+    htmlTable += `<th>${field.name}</th>`;
```

Comment:
I'm not too familiar with jupyter: Can this value be set by the user? If so we should escape it to prevent injection attacks.

Reply:
In this case arbitrary data loaded from a CSV or parquet could emit something nefarious or the more benign -- mess up formatting.

We should definitely escape it, thank you!

Reply:
I've brought in `escape` from https://deno.land/std@0.203.0/html/mod.ts.